src/rosegold/control/inventory.cr

```diff
@@ -75,19 +75,71 @@ class Rosegold::Inventory
   #   inventory.deposit_at_least 5, "diamond_pickaxe" # => 3
   #   inventory.deposit_at_least 5, &.empty?, hotbar # => 1
   #   inventory.deposit_at_least 5, { |slot| slot.item_id == "diamond_pickaxe" && slot.efficiency >= 4 } # => 2
-  def deposit_at_least(count, spec, source : Array(WindowSlot) = inventory + hotbar)
-    shift_click_at_least count, spec, source
+  def deposit_at_least(count, spec, source : Array(WindowSlot) = inventory + hotbar, target : Array(WindowSlot) = content)
+    transferred = 0
+  
+    # prefer large stacks for minimum clicks
+    # for equal stacks, preserve order
+    source.sort_by { |s| s.count }.each do |slot|
+      next unless slot.matches? spec
+      
+      # Find empty slot in target container
+      target_slot = nil
+      target.each_with_index do |slot, index|
+        if slot.matches?("air")
+          target_slot = slot
+        end
+      end
+
+      break if target_slot.nil? # Target full.
+
+      # Swap slots
+      swap_slot = target_slot.slot_nr
+
+      target_slot.slot_nr = slot.slot_nr
+      slot.slot_nr = swap_slot
+
+      changed_slots = [
+       slot,
+       target_slot
+      ]
+
+      client.send_packet! Serverbound::ClickWindow.new :shift, 0_u8, slot.slot_nr.to_u16, changed_slots, client.window.id.to_u8, client.window.state_id, slot
+
+      transferred += slot.count
+
+      break if transferred >= count
+    end
+
+    client.send_packet! Serverbound::CloseWindow.new client.window.id
+
+    transferred
   end
 
   def deposit_at_least(count, &spec : WindowSlot -> _)
     deposit_at_least(count, spec)
   end
 
+  # Closes the current open window if any
+  def close()
+    client.window.close
+  end
+
+  private def find_slot(spec, source) 
+    source.each_with_index do |slot, index|
+      if slot.matches?(spec)
+        return slot
+      end
+    end
+
+    return nil
+  end
+
   private def shift_click_at_least(count, spec, slots : Array(WindowSlot))
     transferred = 0
     # prefer large stacks for minimum clicks
     # for equal stacks, preserve order
-    slots.sort_by { |s| -s.count }.each do |slot|
+    slots.sort_by { |s| s.count }.each do |slot|
```

Comment:
If you decide to start with small stacks (default sort order is ascending), the comment should be changed/removed too.

Reply:
This was probably due to a compilation error because it was trying to change an unsigned to a negative, can fix by doing .to_i8 if you want to retain